example/src/main/java/com/klarna/example/MainActivity.java

```diff
@@ -93,14 +128,86 @@ public void openKlarnaPreferences(View view) {
         Intent intent = new Intent(this, PreferencesActivity.class);
         intent.putExtra(PreferencesActivity.EXTRA_USER_TOKEN, getUserToken());
         startActivityForResult(intent, PREFERENCES_REQUEST_CODE);
+    }
 
+    public void onBuyPressed(View view) {
+        if (this.hasUserToken()) {
+            buyTicket();
+        } else {
+            openKlarnaRegistration();
+        }
     }
+    
+    private void buyTicket() {
+        class purchaseItemRunnable implements Runnable {
+            String reference;
+            OriginProof originProof;
+
+            purchaseItemRunnable(String reference, OriginProof originProof) {
+                this.reference = reference;
+                this.originProof = originProof;
+            }
+
+            @Override
+            public void run() {
+                try {
+                    performPurchaseOfItem(reference, originProof);
+                } catch (final Exception e) {
+                    Handler mainHandler = new Handler(getApplicationContext().getMainLooper());
```

Comment:
How about postToMainHandler(Runnable runnable) function, which we could use also in performPurchaseOfItem